### PR TITLE
fix(query): Avoid race in pipeline delegator

### PR DESCRIPTION
**Pull request by nikhars**
This fixes the race in pipeline delegator by making a copy of the query.
But there is an override mechanism via options wherein we can override
copying query for certain referrers. These would be the
subscription and subscriptions_executor referrers since they are
too sensitive to copying and they would be disallowed from running
parallel queries.

### CODE REVIEW

File: snuba/pipeline/pipeline_delegator.py [hunk 1]
```diff
@@ -26,6 +29,20 @@
 ]
 
 
+def _is_query_copying_disallowed(current_referrer: str) -> bool:
+    """
+    Check if the given referrer is disallowed from making a copy of the query.
```
Comment (MeredithAnya) [inline snuba/pipeline/pipeline_delegator.py:34]: maybe you could add why a given referrer would be disallowed from making a copy of the query? I'm not sure it's fully obvious to me why subscriptions and subscriptions executor are too sensitive for using deepcopy

**Comment by onewland** [comment]
Too sensitive is referring to the increased query latency we saw on the original deploy of that change, right?

**Comment by onewland** [comment]
Is there any way we could do gradual/run-time config roll-out so that it's very straightforward to roll back?

**Review by evanh — APPROVED** [review]
We need a better way of doing this. That's a bigger problem though.

**Pull request merged**